File: apps/gui/src/components/conflict_rows.ts
import { translate, type Locale } from "../i18n/catalog";
import type {
  ConflictAuditScope,
  ConflictBaselineProjection,
  ConflictContentProjection,
  ConflictFileProjection,
  ConflictHunkProjection,
} from "../models/conflict";
import { CONFLICT_CONTENT_CAPABILITY } from "../models/conflict";
import "./conflict_rows.css";

/**
 * The one structured-conflict renderer in the GUI
 * (`runtime.conflict_content`, GUI-CORE-015).
 *
 * It draws what Core published and nothing else: **two sides plus the patch
 * preimage, never a three-way merge.** OURS is a read-only read of the Lane's
 * current file at the hunk's declared old range, THEIRS is the incoming
 * patch's new side, and BASE is the hunk's own preimage — the text the patch
 * expected to find. Core computes no merge base and resolves nothing, so this
 * component renders no merged text and offers no resolve control. That is the
 * whole reason D12 has no manual-merge escape hatch: a hunk resolved in the
 * gate would be code that never passed the Lane's own gates.
 *
 * Rows reuse the registered `.diffbody > .dl` family (`.ln` gutter, `.tx`
 * text) that `gui-kit.css` carries for DiffReview, so a conflict line and a
 * diff line are the same object on screen. Nothing here parses text: line
 * numbers are counted forward from Core's own `oursStart` / `theirsStart`, and
 * a hunk whose start Core published as `0` — the patch declared none — gets no
 * number rather than a fabricated `1`.
 */

/// Rows Core did not publish, each with its own sentence. None of them means
/// "no conflict here", and none may render as an empty body.
type ConflictNoteKind = "omitted" | "no-hunks" | "empty-side";

function note(host: HTMLElement, kind: ConflictNoteKind, text: string): HTMLElement {
  const row = document.createElement("p");
  row.className = "dl note";
  row.dataset.conflictNote = kind;
  row.textContent = text;
  host.append(row);
  return row;
}

/**
 * One side's lines as registered `.dl` rows.
 *
 * `start` is Core's own 1-based first line; `0` means the patch declared none,
 * and then no number is drawn at all. `side` only tints the rows — it never
 * claims a row was added or removed, because a conflict side is neither.
 */
function renderSide(
  host: HTMLElement,
  side: "ours" | "theirs" | "base",
  start: number,
  lines: string[],
  locale: Locale,
): HTMLElement {
  const body = document.createElement("div");
  body.className = `diffbody cc-body cc-${side}`;
  body.dataset.conflictSide = side;
  if (lines.length === 0) {
    note(body, "empty-side", translate(locale, "d12.conflict.emptySide", {}));
    host.append(body);
    return body;
  }
  lines.forEach((line, index) => {
    const row = document.createElement("div");
    row.className = `dl cc-line cc-${side}`;
    const number = document.createElement("span");
    number.className = "ln";
    number.textContent = start > 0 ? String(start + index) : "";
    const text = document.createElement("span");
    text.className = "tx";
    // Core sends the line as it stands in the file, newline included; the row
    // is already `white-space: pre`, so the trailing break would draw a blank
    // line under every row.
    text.textContent = line.replace(/\r?\n$/, "");
    row.append(number, text);
    body.append(row);
  });
  host.append(body);
  return body;
}

/// The five reasons Core names today. `ConflictHunkReason` is open-ended, so
/// anything else keeps Core's own tag rather than borrowing one of these.
const NAMED_REASONS = new Set([
  "context_mismatch",
  "already_applied",
  "file_missing",
  "file_deleted",
  "binary",
]);

function renderReason(host: HTMLElement, reason: string, locale: Locale): void {
  const row = document.createElement("p");
  row.className = "cc-reason";
  const chip = document.createElement("span");
  chip.className = "cc-rchip";
  chip.dataset.conflictReason = reason;
  const remedy = document.createElement("span");
  remedy.className = "cc-remedy";
  remedy.dataset.conflictRemedy = reason;
  if (NAMED_REASONS.has(reason)) {
    // The keys are the exact Core discriminants, so a new reason fails the
    // catalog parity test rather than silently rendering a known sentence.
    chip.textContent = translate(
      locale,
      `d12.conflict.reason.${reason}` as "d12.conflict.reason.context_mismatch",
      {},
    );
    remedy.textContent = translate(
      locale,
      `d12.conflict.reason.${reason}.remedy` as "d12.conflict.reason.context_mismatch.remedy",
      {},
    );
  } else {
    chip.textContent = translate(locale, "d12.conflict.reason.unnamed", { reason });
    remedy.textContent = translate(locale, "d12.conflict.reason.unnamed.remedy", {});
  }
  row.append(chip, remedy);
  host.append(row);
}

function renderHunk(
  host: HTMLElement,
  hunk: ConflictHunkProjection,
  index: number,
  locale: Locale,
): void {
  const block = document.createElement("div");
  block.className = "cc-hunk";
  block.dataset.conflictHunk = String(index);
  renderReason(block, hunk.reason, locale);

  const sides = document.createElement("div");
  sides.className = "cc-sides";
  for (const side of ["ours", "theirs"] as const) {
    const column = document.createElement("div");
    column.className = "cc-col";
    const heading = document.createElement("p");
    heading.className = "cc-shead";
    const start = side === "ours" ? hunk.oursStart : hunk.theirsStart;
    heading.textContent =
      start > 0
        ? translate(locale, side === "ours" ? "d12.conflict.ours" : "d12.conflict.theirs", {
            line: String(start),
          })
        : translate(
            locale,
            side === "ours" ? "d12.conflict.oursUnnumbered" : "d12.conflict.theirsUnnumbered",
            {},
          );
    column.append(heading);
    renderSide(column, side, start, side === "ours" ? hunk.ours : hunk.theirs, locale);
    sides.append(column);
  }
  block.append(sides);

  // The preimage is the third side, collapsed by default: it is what the patch
  // expected, not what either side holds. `null` and `[]` are different facts —
  // "no preimage at all" versus "it expected an empty region" — and get
  // different sentences rather than one shared empty strip.
  if (hunk.base === null) {
    const absent = document.createElement("p");
    absent.className = "dl note cc-baseline-note";
    absent.dataset.conflictBase = "absent";
    absent.textContent = translate(locale, "d12.conflict.base.absent", {});
    block.append(absent);
  } else if (hunk.base.length === 0) {
    const empty = document.createElement("p");
    empty.className = "dl note cc-baseline-note";
    empty.dataset.conflictBase = "empty";
    empty.textContent = translate(locale, "d12.conflict.base.empty", {});
    block.append(empty);
  } else {
    const strip = document.createElement("details");
    strip.className = "cc-base";
    strip.dataset.conflictBase = "present";
    const summary = document.createElement("summary");
    summary.textContent = translate(locale, "d12.conflict.base", {});
    strip.append(summary);
    // The preimage carries the hunk's old-side numbering, the same range OURS
    // was read at, because that is the region the patch expected to find.
    renderSide(strip, "base", hunk.oursStart, hunk.base, locale);
    block.append(strip);
  }

  host.append(block);
}

function renderFile(host: HTMLElement, file: ConflictFileProjection, locale: Locale): void {
  const block = document.createElement("div");
  block.className = "cc-file";
  block.dataset.conflictFile = file.path;
  const heading = document.createElement("p");
  heading.className = "cc-fhead";
  heading.title = file.path;
  heading.textContent = file.path;
  block.append(heading);

  if (file.omitted) {
    // The entry survives on purpose: a reviewer must be able to tell "not
    // shown" from "this file was fine".
    note(block, "omitted", translate(locale, "d12.conflict.omitted", {}));
  } else if (file.hunks.length === 0) {
    // Core listed the file with neither hunks nor the flag. That is not "no
    // conflict" either — it is a file this build cannot account for.
    note(block, "no-hunks", translate(locale, "d12.conflict.noHunks", {}));
  }
  file.hunks.forEach((hunk, index) => renderHunk(block, hunk, index, locale));
  host.append(block);
}

function renderBaseline(
  host: HTMLElement,
  baseline: ConflictBaselineProjection,
  locale: Locale,
  onOpenEvidence?: (scope: ConflictAuditScope) => void,
): void {
  const row = document.createElement("p");
  row.className = "cc-baseline";
  row.dataset.conflictBaseline = baseline.kind;
  const label = document.createElement("span");
  label.className = "cc-blabel";
  label.textContent = translate(locale, "d12.conflict.baseline", {});
  const value = document.createElement("span");
  value.className = "cc-bvalue";
  if (baseline.kind === "revision") {
    value.textContent = translate(locale, "d12.conflict.baseline.revision", {
      sha: baseline.shortSha ?? baseline.sha ?? "",
    });
    if (baseline.sha) value.title = baseline.sha;
  } else if (baseline.kind === "evidence") {
    value.textContent = translate(locale, "d12.conflict.baseline.evidence", {});
  } else if (baseline.kind === "unknown") {
    // A real Core answer: it held no baseline it could name. Never `HEAD`.
    value.textContent = translate(locale, "d12.conflict.baseline.unknown", {});
  } else {
    // `ConflictBaseline` is open-ended; an unmodelled kind reaches the screen
    // as itself rather than collapsing into `unknown`.
    value.textContent = translate(locale, "d12.conflict.baseline.unnamed", {
      kind: baseline.kind,
    });
  }
  row.append(label, value);

  for (const binding of baseline.bindings) {
    // The gate's baseline is its evidence, so each binding is a cross-link.
    // `AuditQuery` filters by object, so the chip carries the audit object
    // Core linked — the same route D12's revert rows already take.
    const chip = document.createElement("button");
    chip.type = "button";
    chip.className = "cc-evchip";
    chip.dataset.conflictEvidence = binding.evidenceId;
    chip.textContent = `${binding.evidenceId} · ${binding.shortHash}`;
    chip.title = `${binding.evidenceId} · ${binding.sourceHash}`;
    chip.setAttribute(
      "aria-label",
      translate(locale, "d12.conflict.openEvidence", { evidence: binding.evidenceId }),
    );
    chip.disabled = onOpenEvidence === undefined;
    if (onOpenEvidence) {
      chip.addEventListener("click", () => onOpenEvidence(binding.auditScope));
    }
    row.append(chip);
  }
  host.append(row);
}

/**
 * Renders one `ConflictContent` into `host`.
 *
 * The "not a merge result" statement is drawn with every pane rather than once
 * per screen, because the pane is what an operator reads and the claim it must
 * not make is exactly the one the layout invites.
 */
export function renderConflictContent(
  host: HTMLElement,
  content: ConflictContentProjection,
  locale: Locale,
  onOpenEvidence?: (scope: ConflictAuditScope) => void,
): HTMLElement {
  const pane = document.createElement("div");
  pane.className = "cc";
  pane.dataset.conflictContent = "true";

  const claim = document.createElement("p");
  claim.className = "cc-notmerge";
  claim.dataset.conflictNotMerge = "true";
  claim.textContent = translate(locale, "d12.conflict.notMerge", {});
  pane.append(claim);

  renderBaseline(pane, content.baseline, locale, onOpenEvidence);

  if (content.truncated) {
    const banner = document.createElement("p");
    banner.className = "cc-trunc";
    banner.dataset.conflictTruncated = "true";
    banner.setAttribute("role", "status");
    banner.textContent = translate(locale, "d12.conflict.truncated", {});
    pane.append(banner);
  }

  if (content.files.length === 0) {
    // Core published content naming no file. That is not "no conflict"; it is
    // content this build cannot account for.
    note(pane, "no-hunks", translate(locale, "d12.conflict.noFiles", {}));
  }
  for (const file of content.files) {
    renderFile(pane, file, locale);
  }

  host.append(pane);
  return pane;
}

/**
 * The two ways conflict lines can be absent, which are different facts.
 *
 * `capabilityAvailable: false` is a Core that publishes no structured content
 * at all — the reason text is everything there is. With the capability
 * present, an absent content means Core published none for *this* record: an
 * operator `BounceMergeConflict` is a human judgement with no failed apply
 * behind it and carries none by contract. Neither ever means the conflict was
 * empty.
 */
export function renderConflictUnavailable(
  host: HTMLElement,
  locale: Locale,
  capabilityAvailable: boolean,
): HTMLElement {
  const row = document.createElement("p");
  row.className = "cc-absent";
  if (capabilityAvailable) {
    row.dataset.conflictAbsent = "record";
    row.textContent = translate(locale, "d12.conflict.absent.record", {});
  } else {
    row.dataset.conflictAbsent = "capability";
    row.textContent = translate(locale, "d12.conflict.absent.capability", {
      capability: CONFLICT_CONTENT_CAPABILITY,
    });
  }
  host.append(row);
  return row;
}

File: apps/tui/src/tui/conflict_rows.rs
//! Structured conflict rows (`runtime.conflict_content`, GUI-CORE-015).
//!
//! A failed merge or lane apply used to publish one prose sentence, so a client
//! could say *that* a patch conflicted and never *what* conflicted. Core now
//! publishes the lines. This module renders them and does nothing else.
//!
//! The invariant that shapes every row here: **this is not a three-way merge.**
//! `ours` is a read-only read of the target file where the hunk said its region
//! was, `theirs` is the incoming hunk's new side, and `base` is that hunk's own
//! preimage — the text the patch expected to find. Core computes no merge base
//! and resolves nothing, so this surface renders three sides, says so in words,
//! and never offers to auto-resolve.
//!
//! Two absences are different facts and are rendered differently: `base: None`
//! means the hunk had no preimage at all, while `Some(vec![])` means it
//! expected an empty region, which is what a creation hunk expects. Collapsing
//! them would tell a reviewer the patch expected nothing when it expected an
//! empty file.

use viden_core::{
    ConflictBaseline, ConflictContent, ConflictFile, ConflictHunk, ConflictHunkReason,
};

use super::{state::TuiState, text::truncate_end, text::truncate_tail};

/// The capability that publishes `ConflictBounce.content` and
/// `LaneConflictView.content`.
pub(super) const CONFLICT_CONTENT_CAPABILITY: &str = "runtime.conflict_content";

/// Largest number of detail rows one conflict modal renders.
///
/// A 256 KiB payload would otherwise scroll past anything an operator can read;
/// the remainder is counted in a stated row rather than dropped in silence.
pub(super) const MAX_CONFLICT_DETAIL_ROWS: usize = 60;

/// The countable facts of one `ConflictContent`, for a one-line row.
///
/// A projection of Core's own payload — counts and the baseline kind — not a
/// second copy of the lines. It exists so the Decision Center can say how big
/// a conflict is without cloning a quarter-megabyte body into every frame.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(super) struct ConflictContentSummary {
    pub(super) files: usize,
    pub(super) hunks: usize,
    pub(super) omitted_files: usize,
    pub(super) truncated: bool,
    pub(super) baseline: ConflictBaselineTag,
}

/// What the `ours` side was read against, as a short tag.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(super) enum ConflictBaselineTag {
    Revision,
    Evidence,
    Unknown,
    /// A baseline kind a newer Core published that this build cannot name.
    /// Reported as unrecognized, never silently as `HEAD`.
    Unrecognized,
}

impl ConflictBaselineTag {
    pub(super) const fn label_key(self) -> &'static str {
        match self {
            Self::Revision => "conflict.baseline.revision",
            Self::Evidence => "conflict.baseline.evidence",
            Self::Unknown => "conflict.baseline.unknown",
            Self::Unrecognized => "conflict.baseline.unrecognized",
        }
    }
}

impl From<&ConflictBaseline> for ConflictBaselineTag {
    fn from(baseline: &ConflictBaseline) -> Self {
        match baseline {
            ConflictBaseline::Revision { .. } => Self::Revision,
            ConflictBaseline::Evidence { .. } => Self::Evidence,
            ConflictBaseline::Unknown => Self::Unknown,
            _ => Self::Unrecognized,
        }
    }
}

impl From<&ConflictContent> for ConflictContentSummary {
    fn from(content: &ConflictContent) -> Self {
        Self {
            files: content.files.len(),
            hunks: content.files.iter().map(|file| file.hunks.len()).sum(),
            omitted_files: content.files.iter().filter(|file| file.omitted).count(),
            truncated: content.truncated,
            baseline: ConflictBaselineTag::from(&content.baseline),
        }
    }
}

/// The one-line conflict fact for a decision row or a transcript entry.
///
/// `None` when the capability is absent or Core published no content, so the
/// caller keeps today's reason-only row. `None` means Core has nothing to show,
/// never that the conflict was empty.
pub(super) fn content_summary_row(
    state: &TuiState,
    summary: Option<ConflictContentSummary>,
) -> Option<String> {
    if !state.has_capability(CONFLICT_CONTENT_CAPABILITY) {
        return None;
    }
    let summary = summary?;
    let mut row = super::i18n::translate(
        state,
        "conflict.summary",
        &[
            ("files", &summary.files.to_string()),
            ("hunks", &summary.hunks.to_string()),
            (
                "baseline",
                &super::i18n::text(state, summary.baseline.label_key()),
            ),
        ],
    );
    if summary.omitted_files > 0 {
        row.push_str(&super::i18n::translate(
            state,
            "conflict.summary.omitted",
            &[("count", &summary.omitted_files.to_string())],
        ));
    }
    if summary.truncated {
        row.push_str(&super::i18n::text(state, "conflict.summary.truncated"));
    }
    Some(row)
}

/// The full detail body: three labelled sides per rejected hunk.
pub(super) fn detail_rows(
    state: &TuiState,
    content: Option<&ConflictContent>,
    width: usize,
) -> Vec<String> {
    if !state.has_capability(CONFLICT_CONTENT_CAPABILITY) {
        return vec![super::i18n::translate(
            state,
            "conflict.detail.unavailable",
            &[("capability", CONFLICT_CONTENT_CAPABILITY)],
        )];
    }
    let Some(content) = content else {
        return vec![super::i18n::text(state, "conflict.detail.absent")];
    };
    // The disclaimer is the first row, before any line an operator could read
    // as a resolution. It is mandatory: Core resolved nothing.
    let mut rows = vec![
        super::i18n::translate(
            state,
            "conflict.detail.baseline",
            &[(
                "baseline",
                &super::i18n::text(
                    state,
                    ConflictBaselineTag::from(&content.baseline).label_key(),
                ),
            )],
        ),
        super::i18n::text(state, "conflict.detail.not_a_merge"),
    ];
    let mut body = Vec::new();
    for file in &content.files {
        body.extend(file_rows(state, file, width));
    }
    let hidden = body.len().saturating_sub(MAX_CONFLICT_DETAIL_ROWS);
    if hidden > 0 {
        body.truncate(MAX_CONFLICT_DETAIL_ROWS);
        body.push(super::i18n::translate(
            state,
            "conflict.detail.more",
            &[("count", &hidden.to_string())],
        ));
    }
    rows.extend(body);
    if content.truncated {
        rows.push(super::i18n::text(state, "conflict.detail.truncated"));
    }
    rows.into_iter()
        .map(|row| truncate_end(&row, width))
        .collect()
}

fn file_rows(state: &TuiState, file: &ConflictFile, width: usize) -> Vec<String> {
    let mut rows = vec![super::i18n::translate(
        state,
        "conflict.detail.file",
        &[("path", &truncate_tail(&file.path, width.saturating_sub(4)))],
    )];
    if file.omitted {
        rows.push(super::i18n::text(state, "conflict.detail.omitted"));
    }
    for hunk in &file.hunks {
        rows.extend(hunk_rows(state, hunk, width));
    }
    rows
}

fn hunk_rows(state: &TuiState, hunk: &ConflictHunk, width: usize) -> Vec<String> {
    let mut rows = vec![super::i18n::translate(
        state,
        "conflict.detail.hunk",
        &[(
            "reason",
            &super::i18n::text(state, hunk_reason_key(hunk.reason)),
        )],
    )];
    rows.extend(side_rows(
        state,
        "conflict.side.ours",
        hunk.ours_start,
        &hunk.ours,
        width,
    ));
    rows.extend(side_rows(
        state,
        "conflict.side.theirs",
        hunk.theirs_start,
        &hunk.theirs,
        width,
    ));
    match hunk.base.as_deref() {
        // The patch expected an empty region, which is what a creation hunk
        // expects. Distinct from "there was no preimage at all".
        Some([]) => rows.push(super::i18n::text(state, "conflict.side.base_empty")),
        Some(lines) => rows.extend(side_rows(
            state,
            "conflict.side.base",
            // The preimage is the patch's own text; it carries no separate
            // start, so the side is labelled without a line number rather than
            // borrowing one from another side.
            0,
            lines,
            width,
        )),
        None => rows.push(super::i18n::text(state, "conflict.side.base_absent")),
    }
    rows
}

/// One labelled side. An empty side is stated, because "the file has nothing
/// there" is a fact a reviewer needs, not a row to drop.
fn side_rows(
    state: &TuiState,
    side_key: &str,
    start: u32,
    lines: &[String],
    width: usize,
) -> Vec<String> {
    let side = super::i18n::text(state, side_key);
    if lines.is_empty() {
        return vec![super::i18n::translate(
            state,
            "conflict.side.empty",
            &[("side", &side)],
        )];
    }
    lines
        .iter()
        .enumerate()
        .map(|(offset, line)| {
            let number = if start == 0 {
                String::new()
            } else {
                (start + offset as u32).to_string()
            };
            format!(
                "  {side:<7}{number:>5}  {}",
                truncate_end(
                    line.trim_end_matches(['\n', '\r']),
                    width.saturating_sub(16)
                )
            )
        })
        .collect()
}

const fn hunk_reason_key(reason: ConflictHunkReason) -> &'static str {
    match reason {
        ConflictHunkReason::ContextMismatch => "conflict.reason.context_mismatch",
        ConflictHunkReason::AlreadyApplied => "conflict.reason.already_applied",
        ConflictHunkReason::FileMissing => "conflict.reason.file_missing",
        ConflictHunkReason::FileDeleted => "conflict.reason.file_deleted",
        ConflictHunkReason::Binary => "conflict.reason.binary",
        // `#[non_exhaustive]`: a rejection a newer apply path saw is named as
        // unrecognized rather than mislabelled as a context mismatch.
        _ => "conflict.reason.unrecognized",
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use viden_core::CapabilityId;

    fn capable() -> TuiState {
        let mut state = TuiState::default();
        state
            .capabilities
            .insert(CapabilityId(CONFLICT_CONTENT_CAPABILITY.to_string()));
        state
    }

    fn hunk(base: Option<Vec<String>>) -> ConflictHunk {
        ConflictHunk {
            ours_start: 42,
            ours: vec!["    let bounce = record_conflict_bounce(gate)?;\n".to_string()],
            theirs_start: 42,
            theirs: vec!["    let bounce = bounce_with_reason(gate, reason)?;\n".to_string()],
            base,
            reason: ConflictHunkReason::ContextMismatch,
        }
    }

    fn content(base: Option<Vec<String>>) -> ConflictContent {
        ConflictContent {
            baseline: ConflictBaseline::Unknown,
            files: vec![ConflictFile {
                path: "crates/runtime/src/trust_loop.rs".to_string(),
                hunks: vec![hunk(base)],
                omitted: false,
            }],
            truncated: false,
        }
    }

    #[test]
    fn the_detail_always_says_it_is_not_a_merge_result() {
        let rows = detail_rows(
            &capable(),
            Some(&content(Some(vec![
                "    let bounce = record_bounce(gate)?;".to_string(),
            ]))),
            96,
        )
        .join("\n");

        assert!(rows.contains("not a merge result"), "{rows}");
        assert!(rows.contains("OURS"), "{rows}");
        assert!(rows.contains("THEIRS"), "{rows}");
        assert!(rows.contains("BASE"), "{rows}");
        // Core's own line numbers, on the sides that have them.
        assert!(rows.contains("42"), "{rows}");
        assert!(rows.contains("context mismatch"), "{rows}");
    }

    /// `None` and `Some(vec![])` are different facts about the patch preimage.
    #[test]
    fn an_absent_preimage_and_an_empty_one_never_render_alike() {
        let state = capable();
        let absent = detail_rows(&state, Some(&content(None)), 96).join("\n");
        let empty = detail_rows(&state, Some(&content(Some(Vec::new()))), 96).join("\n");

        assert_ne!(absent, empty);
        assert!(absent.contains("no preimage"), "{absent}");
        assert!(empty.contains("empty region"), "{empty}");
    }

    #[test]
    fn absent_content_and_a_missing_capability_are_stated_separately() {
        let capable = capable();
        assert!(
            detail_rows(&capable, None, 96)[0].contains("published no content"),
            "{:?}",
            detail_rows(&capable, None, 96)
        );
        let base = TuiState::default();
        assert!(
            detail_rows(&base, Some(&content(None)), 96)[0].contains(CONFLICT_CONTENT_CAPABILITY)
        );
        // Without the capability there is no summary row either, so the caller
        // keeps the reason-only row it always had.
        assert!(
            content_summary_row(&base, Some(ConflictContentSummary::from(&content(None))))
                .is_none()
        );
    }

    #[test]
    fn the_summary_counts_files_hunks_omissions_and_names_the_baseline() {
        let state = capable();
        let mut value = content(None);
        value.files.push(ConflictFile {
            path: "crates/runtime/src/huge.rs".to_string(),
            hunks: Vec::new(),
            omitted: true,
        });
        value.truncated = true;
        value.baseline = ConflictBaseline::Revision {
            sha: "9f9f9f9f".to_string(),
        };

        let summary = ConflictContentSummary::from(&value);
        assert_eq!(summary.files, 2);
        assert_eq!(summary.hunks, 1);
        assert_eq!(summary.omitted_files, 1);
        assert_eq!(summary.baseline, ConflictBaselineTag::Revision);

        let row = content_summary_row(&state, Some(summary)).expect("summary row");
        assert!(row.contains('2') && row.contains('1'), "{row}");
        assert!(row.contains("revision"), "{row}");
        assert!(row.contains("omitted"), "{row}");
        assert!(row.contains("truncated"), "{row}");
    }

    #[test]
    fn every_row_fits_the_requested_width() {
        let state = capable();
        let value = content(Some(vec![
            "    let bounce = record_bounce(gate)?; // a very long trailing comment".to_string(),
        ]));
        for width in [32usize, 60, 96, 132] {
            for row in detail_rows(&state, Some(&value), width) {
                assert!(
                    super::super::text::char_width(&row) <= width,
                    "width {width} overflowed: {row}"
                );
            }
        }
    }
}
